### PR TITLE
V1.4: Storage-Testing

### PR DESCRIPTION
**Pull request by alson001**
toModelType throws IllegalArgumentException instead of IllegalValueException cause the Ingredient checkArguments from App.util return IllegalArgumentException and this is not surrounded by try catch to return new IllegalValueException

### CODE REVIEW

**Review by wgw0 — COMMENTED** [review]
## Performance optimising changes
1. In the `toModelType_validIngredientDetails_returnsIngredientKeyValuePair()` method, the `build()` method can be refactored to return a list of `HashMap<Ingredient, IngredientInformation>` objects to improve performance.

2. In the `toModelType_validIngredientDetails_returnsIngredientKeyValuePair()` method, a single `JsonAdaptedIngredient` object can be created and reused for each `Ingredient` in the loop to improve performance.

3. In the `toModelType_validIngredientDetails_returnsIngredientKeyValuePair()` method, a `HashMap<Ingredient, IngredientInformation>` may be inefficient in terms of memory usage and time complexity, and a more memory-efficient data structure can be used instead, such as a `TreeMap` or an array, if the `Ingredient` objects are sorted.

4. In the `toModelType_validIngredientDetails_returnsIngredientKeyValuePair()` method, catching an `IllegalValueException` and throwing a new `RuntimeException` can be inefficient and make it difficult to diagnose errors. A better approach would be to log the error instead of throwing a new exception, or to handle the error in a more appropriate way.

5. In the `toModelType_validIngredientDetails_returnsIngredientKeyValuePair()` method, calling `putAll()` on the `ingredientList` `HashMap` for each `JsonAdaptedIngredient` object may be inefficient if the `JsonAdaptedIngredient` objects contain the same keys, as the `putAll()` method will overwrite existing values. A better approach would be to use the `put()` method instead of `putAll()`, or to check for duplicate keys before calling `putAll()`.

**Review by wgw0 — APPROVED** [review]

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedIngredientTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                        new JsonAdaptedIngredient(ingredient, ingredientInformation);
+                HashMap<Ingredient, IngredientInformation> i = null;
+                try {
+                    i = jsonAdaptedIngredient.toModelType();
+                } catch (IllegalValueException e) {
+                    throw new RuntimeException(e.getMessage());
+                }
+                ingredientList.putAll(i);
+            }));
+        }
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedIngredientTest.java:50]:
```java
CACIO_INGREDIENTS.forEach(ingredientBuilder ->
    ingredientBuilder.build() //Ingredient Map
        .forEach((mainIngredient, ingredientInformation) ->
            try {
                new JsonAdaptedIngredient(mainIngredient, ingredientInformation)
                    .toModelType()
                    .forEach(ingredientList::put);
            } catch (IllegalValueException e) {
                fail(e.getMessage());
            }
        )
);
```

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedNameTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.junit.jupiter.api.Test;
+
+import seedu.recipe.commons.exceptions.IllegalValueException;
+import seedu.recipe.model.recipe.Name;
+
+
+public class JsonAdaptedNameTest {
+
+    private static final String VALID_NAME = "Cheeseburger";
+    private static final String INVALID_NAME = "Burger123";
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedNameTest.java:17]: Might wanna check if they changed the regex to make it more lenient @filbertphang @alson001

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedPortionUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+//import static seedu.recipe.testutil.Assert.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.recipe.commons.exceptions.IllegalValueException;
+import seedu.recipe.model.recipe.unit.PortionUnit;
+
+public class JsonAdaptedPortionUnitTest {
+    private static final String VALID_UNIT = "grams";
+    private static final String INVALID_UNIT = "123";
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedPortionUnitTest.java:13]: Same wrt this @filbertphang @alson001

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedRecipeDurationTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import seedu.recipe.commons.exceptions.IllegalValueException;
+import seedu.recipe.model.recipe.RecipeDuration;
+import seedu.recipe.model.recipe.unit.TimeUnit;
+
+public class JsonAdaptedRecipeDurationTest {
+    private static final RecipeDuration RECIPE_DURATION = new RecipeDuration(30, new TimeUnit("minutes"));
+    private static final double TIME = 30;
+    private static final TimeUnit TIME_UNIT = new TimeUnit("minutes");
+    private static final JsonAdaptedTimeUnit JSON_ADAPTED_TIME_UNIT = new JsonAdaptedTimeUnit(TIME_UNIT);
+
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedRecipeDurationTest.java:18]: Accept plural or non plural for `1` here? @filbertphang @alson001 

i.e. `1 minute` or `1 minutes`

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedStepTest.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import seedu.recipe.commons.exceptions.IllegalValueException;
+import seedu.recipe.model.recipe.Step;
+
+public class JsonAdaptedStepTest {
+
+    private static final String VALID_STEP = "Preheat oven to 200C";
+    private static final String INVALID_STEP = "123456";
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedStepTest.java:14]: Are there more concrete examples of invalid Steps you should assert?

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedSubstitutionIngredientTest.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.recipe.model.recipe.ingredient.Ingredient;
+
+public class JsonAdaptedSubstitutionIngredientTest {
+    private static final String VALID_INGREDIENT_NAME = "Sugar";
+    private static final String INVALID_INGREDIENT_NAME = "";
+
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedSubstitutionIngredientTest.java:16]: Are there other invalid Ingredient Names?

Comment (SeeuSim) [reply src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedSubstitutionIngredientTest.java:16]: Can consider making them a static import from `Ingredient` class @alson001 @ZHTang29

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedTagTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import seedu.recipe.commons.exceptions.IllegalValueException;
+import seedu.recipe.model.tag.Tag;
+
+public class JsonAdaptedTagTest {
+    private static final String VALID_TAG_NAME = "healthy";
+    private static final String INVALID_TAG_NAME = "#fatty";
+
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedTagTest.java:14]: Are there more invalid tags? i.e. whitespace, special characters, leading trailing whitespace, empty

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedTimeUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import seedu.recipe.commons.exceptions.IllegalValueException;
+import seedu.recipe.model.recipe.unit.TimeUnit;
+
+public class JsonAdaptedTimeUnitTest {
+    private static final String VALID_UNIT = "hour";
+    private static final String INVALID_UNIT = "123";
+
```
Comment (SeeuSim) [inline src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedTimeUnitTest.java:15]: Does this cover all equivalence partitions?

Comment (alson001) [reply src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedTimeUnitTest.java:15]: This should be done in model tests.

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedTagTest.java (hunk 7 above)
Comment (alson001) [reply src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedTagTest.java:14]: This should be done in model tests.

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedSubstitutionIngredientTest.java (hunk 6 above)
Comment (alson001) [reply src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedSubstitutionIngredientTest.java:16]: This should be done in model tests.

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedNameTest.java (hunk 2 above)
Comment (alson001) [reply src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedNameTest.java:17]: Changed according to filbert changes.

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedRecipeDurationTest.java (hunk 4 above)
Comment (alson001) [reply src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedRecipeDurationTest.java:18]: Realised no way for it throw IllegalValueException as the check occurs earlier. This plural or non plural should be done in model test.

File: src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedPortionUnitTest.java (hunk 3 above)
Comment (alson001) [reply src/test/java/seedu/recipe/storage/jsonadapters/JsonAdaptedPortionUnitTest.java:13]: Realised no way for it throw IllegalValueException as the check occurs earlier.

**Review by SeeuSim — APPROVED** [review]

**Pull request merged**